dp_check/dp_check.go

```diff
@@ -246,42 +259,73 @@ func main() {
 		}
 		req.Header.Add("Metadata-Flavor", "Google")
 		resp, err := client.Do(req)
-		infoLog.Printf("Check connection to metadata server by sending http GET request to %s", metadataServerUrl)
 		if err != nil {
 			return err
 		}
-		if resp.StatusCode != 200 {
-			return fmt.Errorf(`Connecting to metadata server failed. Either this VM doesn't have DirectPath access, or there is a bug that may be causing a larger outage`)
+		defer resp.Body.Close()
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return err
 		}
-		return nil
-	})
+		IPv6FromMetadataServer = net.ParseIP(strings.TrimSuffix(string(body), "\n"))
 
-	// Check IPv6
-	runCheck("IPv6 routes", func() error {
-		var out string
-		var err error
-		cmd := "ip -6 route show"
-		infoLog.Printf("Check IPv6 routes with:|%v|...\n", cmd)
-		if out, err = getOutput(cmd); err != nil {
+		infoLog.Printf("Check if this VM enables DirectPath by sending http GET request to %s", metadataServerUrl)
+		if err != nil {
 			return err
 		}
-		if !strings.Contains(out, "2001:4860:8040::/42") {
-			return fmt.Errorf(`Missing route prefix to backends: 2001:4860:8040::/42.
-IPv6 route setup either failed or hasn't been attempted`)
+		if resp.StatusCode == 200 {
+			return nil
 		}
-		return nil
+		if resp.StatusCode == 404 {
+			return fmt.Errorf("This VM doesn't have DirectPath access")
+		}
+		return fmt.Errorf("Connecting to metadata server failed. There may be a bug that may be causing a larger outage")
 	})
+
 	runCheck("IPv6 addresses", func() error {
-		var out string
-		var err error
-		cmd := "ip -6 addr show"
-		infoLog.Printf("Check IPv6 addresses with:|%v|...\n", cmd)
-		if out, err = getOutput(cmd); err != nil {
+		ifaces, err := net.Interfaces()
+		if err != nil {
 			return err
 		}
-		if !strings.Contains(out, "inet6 2600") {
+		// Go through all interfaces on the VM to see if IPv6 is enabled and if there is an IPv6 address, then check against the IPv6 address returned from metadataserver
+		var hasIPv6 bool
+		for _, iface := range ifaces {
+			if iface.Flags&net.FlagLoopback != 0 {
+				continue
+			}
+			if iface.Flags&net.FlagUp != net.FlagUp {
+				continue
+			}
+			ifaddrs, err := iface.Addrs()
+			if err != nil {
+				return err
+			}
+			for _, ifaddr := range ifaddrs {
+				ip := ifaddr.(*net.IPNet).IP
+				if ip.To4() == nil {
+					hasIPv6 = true
+					if ip.Equal(IPv6FromMetadataServer) {
+						eth0 = iface
+						return nil
+					}
+				}
+			}
+		}
+		if !hasIPv6 {
```

Comment:
nit: now that the semantics of this test have changed, can we add details to the error message:

something like:

```
This VM was expected to have an network interface with IPv6 address: %v assigned to it, but no such interface was found. IPv6 DHCP setup either failed or hasn't been attempted.
```

Reply:
There are 2 checks here essentially: 1) if ipv6 is enabled at all on the machine 2) if it is then is it the same as what metadata returns.